File: message-bus/src/main/python/adapt/messagebus/rabbitmq/MessageArchiver.py
import json

from threading import Thread

from adapt.messagebus.Envelope import State
from adapt.messagebus.SimpleEnvelope import SimpleEnvelope

from adapt.messagebus.rabbitmq.PersistedEnvelope import PersistedEnvelope

from adapt.messagebus.configuration.PersistedEnvelopeRepository import PersistedEnvelopeRepository

class MessageArchiver(Thread):
    def __init__(self, channel, exchangeName, envelopeRepository, mpq, msrk):
        super().__init__()
        self.__channel = channel
        self.__mpq = mpq
        self.__msrk = msrk
        self.__channel.queue_bind(self.__mpq, exchangeName, "#")
        self.__envelopeRepository = envelopeRepository
        #self.__consumer = MessageLogConsumer(channel, envelopeRepository)
    def handleDelivery(self, consumerTag, envelope, properties, body):
        message = str(body)
        msg = "delivered: " + envelope.routing_key + ": " + message
        #print(msg)
        persistedEnvelope = None
        if envelope.routing_key == self.__msrk:
            stateMessage = str(body)
            t = stateMessage.split("@")
            _id = t[0]
            state = t[1]
            timestamp = t[2]
            
            opt = self.__envelopeRepository.findByEnvelopeId(_id)
            persistedEnvelope = PersistedEnvelopeRepository() if opt is None else opt.get()
            state = State[state.split(".")[1]]
            persistedEnvelope.state = state
            if state == State.Sent:
                persistedEnvelope.sentTimestamp = timestamp
            elif state == State.Received:
                persistedEnvelope.receivedTimestamp = timestamp
            elif state == State.Consumed:
                persistedEnvelope.consumedTimestamp = timestamp
            elif state == State.Expired:
                persistedEnvelope.expiredTimestamp = timestamp
            elif state == State.Rejected:
                persistedEnvelope.rejectedTimestamp = timestamp
            else:
                raise Exception( "Unknown state: '" + state + "'" )
        else:
            simpleEnvelope = SimpleEnvelope()
            if body is not None:
                jsons = str(body, "UTF-8")
                simpleEnvelope.deserialize(json.loads(jsons))
            persistedEnvelope = PersistedEnvelope.convert(simpleEnvelope)
        self.__envelopeRepository.save(persistedEnvelope)
    def run(self):
        autoAck = True
        self.__channel.basic_consume(self.handleDelivery, self.__mpq, autoAck)
        self.__channel.start_consuming()
    def join(self):
        self.__channel.stop_consuming()
        super().join()

File: message-bus/src/main/java/alma/icd/adapt/messagebus/rabbitmq/MessageArchiver.java
package alma.icd.adapt.messagebus.rabbitmq;

import static alma.icd.adapt.messagebus.rabbitmq.RabbitMqMessageBroker.MESSAGE_PERSISTENCE_QUEUE;
import static alma.icd.adapt.messagebus.rabbitmq.RabbitMqMessageBroker.MESSAGE_STATE_ROUTING_KEY;

import java.io.IOException;
import java.util.Optional;
import java.util.concurrent.TimeoutException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.rabbitmq.client.AMQP.BasicProperties;

import alma.icd.adapt.messagebus.SimpleEnvelope;
import alma.icd.adapt.messagebus.configuration.PersistedEnvelopeRepository;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Consumer;
import com.rabbitmq.client.DefaultConsumer;

public class MessageArchiver implements Runnable {
	
	private Logger logger = LoggerFactory.getLogger( MessageArchiver.class );
	private Consumer consumer;
	private Channel channel;
	
	class MessageLogConsumer extends DefaultConsumer {

		private PersistedEnvelopeRepository envelopeRepository;
		public MessageLogConsumer( Channel channel, PersistedEnvelopeRepository envelopeRepository ) {
			super(channel);
			this.envelopeRepository = envelopeRepository;
		}
		
		@Override
		public void handleDelivery( String consumerTag, 
									com.rabbitmq.client.Envelope envelope,
									BasicProperties properties, 
									byte[] body ) throws IOException {

			String message = new String( body );
			String msg = "delivered: " + envelope.getRoutingKey() + ": " + message; 
			logger.info( msg );
			
			PersistedEnvelope persistedEnvelope;
			
			if( envelope.getRoutingKey().equals( MESSAGE_STATE_ROUTING_KEY )) {
				
				String stateMessage = new String( body );
				String[] t = stateMessage.split( "@" );
				String id = t[0];
				String state = t[1];
				String timestamp = t[2];
				
//				Iterable<PersistedEnvelope> all = envelopeRepository.findAll();
//				for (PersistedEnvelope pe : all) {
//					System.out.println( ">>> " + pe );
//				}
				
				Optional<PersistedEnvelope> opt = envelopeRepository.findByEnvelopeId( id );
				persistedEnvelope = opt.get();
				persistedEnvelope.state = state;
				switch( state ) {
				
				case "Sent":
					persistedEnvelope.sentTimestamp = timestamp;
					break;
					
				case "Received":
					persistedEnvelope.receivedTimestamp = timestamp;
					break;
					
				case "Consumed":
					persistedEnvelope.consumedTimestamp = timestamp;
					break;
					
				case "Expired":
					persistedEnvelope.expiredTimestamp = timestamp;
					break;
					
				case "Rejected":
					persistedEnvelope.rejectedTimestamp = timestamp;
					break;
				
				default:
					throw new RuntimeException( "Unknown state: '" + state + "'" );
				}
				
			}
			else {
				String json = new String(body, "UTF-8");
//				System.out.println( ">>>> delivered json: " + json );
				
				ObjectMapper objectMapper = new ObjectMapper();
				SimpleEnvelope simpleEnvelope = objectMapper.readValue( json, SimpleEnvelope.class );
				persistedEnvelope = PersistedEnvelope.convert( simpleEnvelope );
			}
			envelopeRepository.save( persistedEnvelope );

//			Iterable<PersistedEnvelope> all = envelopeRepository.findAll();
//			for (PersistedEnvelope pe : all) {
//				System.out.println( ">>> pe: " + pe );
//			}
		}
	}
	
	public MessageArchiver( Channel channel, 
						   String exchangeName,
						   PersistedEnvelopeRepository envelopeRepository ) 
			throws IOException, TimeoutException {
		
		this.channel = channel;
		this.channel.queueBind( MESSAGE_PERSISTENCE_QUEUE, exchangeName, "#" );
		this.consumer = new MessageLogConsumer( channel, envelopeRepository );
	}
	
	@Override
	public void run() {
		try {
			boolean autoAck = true;
			this.channel.basicConsume( MESSAGE_PERSISTENCE_QUEUE, autoAck, this.consumer );
		} 
		catch( IOException e ) {
			throw new RuntimeException( e );
		}
	}
}